File: frontend/lib/api/research.ts
import { api } from "./client";

export interface ResearchJob {
  id: string;
  business_id: string;
  user_id?: string | null;
  status: "PENDING" | "RUNNING" | "COMPLETED" | "PARTIAL" | "FAILED" | "CANCELLED";
  requested_sections: string[];
  source_count: number;
  records_found: number;
  records_validated: number;
  records_rejected: number;
  started_at?: string | null;
  completed_at?: string | null;
  error_message?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ResearchRecord {
  id: string;
  business_id: string;
  research_job_id?: string | null;
  source_url?: string | null;
  source_trust: "OFFICIAL" | "HIGH_TRUST" | "MEDIUM_TRUST" | "LOW_TRUST" | "UNKNOWN";
  research_type:
    | "IDENTITY"
    | "CONTACT"
    | "LOCATION"
    | "SERVICES"
    | "WEBSITE"
    | "SOCIAL"
    | "HOURS"
    | "DESCRIPTION"
    | "PUBLIC_REVIEWS"
    | "GENERAL";
  field_name: string;
  raw_value?: string | null;
  normalized_value: string;
  confidence: "HIGH" | "MEDIUM" | "LOW";
  evidence_text?: string | null;
  observed_at: string;
  expires_at?: string | null;
  status: "VALIDATED" | "REJECTED" | "STALE" | "CONFLICT";
  meta_info?: Record<string, any> | null;
  created_at: string;
}

export interface ResearchConflict {
  id: string;
  business_id: string;
  field_name: string;
  competing_values: Array<{
    value: string;
    source_url?: string | null;
    confidence: string;
    observed_at?: string | null;
  }>;
  status: "NO_CONFLICT" | "CONFLICT" | "RESOLVED" | "REQUIRES_REVIEW";
  resolution_notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface BusinessResearchProfile {
  business_id: string;
  business_name: string;
  total_jobs: number;
  total_records: number;
  active_conflicts_count: number;
  last_researched_at?: string | null;
  confidence_score: number;
  records: ResearchRecord[];
  conflicts: ResearchConflict[];
}

export interface ResearchJobCreateInput {
  business_id: string;
  sections?: string[];
  provider_type?: "MOCK" | "WEB";
}

export async function createResearchJob(data: ResearchJobCreateInput): Promise<ResearchJob> {
  const response = await api<{ data: ResearchJob }>("/research/jobs", {
    method: "POST",
    body: JSON.stringify(data),
  });
  return response.data;
}

export async function listResearchJobs(params: {
  page?: number;
  page_size?: number;
  business_id?: string;
  status?: string;
} = {}): Promise<{ data: ResearchJob[]; total: number }> {
  const query = new URLSearchParams();
  if (params.page) query.set("page", params.page.toString());
  if (params.page_size) query.set("page_size", params.page_size.toString());
  if (params.business_id) query.set("business_id", params.business_id);
  if (params.status) query.set("status", params.status);

  const response = await api<{ data: ResearchJob[]; pagination: { total: number } }>(
    `/research/jobs?${query.toString()}`
  );
  return { data: response.data, total: response.pagination.total };
}

export async function runResearchJob(jobId: string, providerType: string = "MOCK"): Promise<ResearchJob> {
  const response = await api<{ data: ResearchJob }>(`/research/jobs/${jobId}/run?provider_type=${providerType}`, {
    method: "POST",
  });
  return response.data;
}

export async function cancelResearchJob(jobId: string): Promise<ResearchJob> {
  const response = await api<{ data: ResearchJob }>(`/research/jobs/${jobId}/cancel`, {
    method: "POST",
  });
  return response.data;
}

export async function getBusinessResearchProfile(businessId: string): Promise<BusinessResearchProfile> {
  const response = await api<{ data: BusinessResearchProfile }>(`/research/businesses/${businessId}`);
  return response.data;
}

export async function runResearchAgentJob(jobId: string): Promise<ResearchJob> {
  const response = await api<{ data: ResearchJob }>(`/research/jobs/${jobId}/run-agent`, {
    method: "POST",
  });
  return response.data;
}

File: backend/app/schemas/research.py
import uuid
from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ResearchJobCreate(BaseModel):
    business_id: uuid.UUID
    sections: List[str] = Field(default_factory=lambda: ["ALL"])
    provider_type: str = Field(default="MOCK")  # "MOCK" or "WEB"


class ResearchJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    business_id: uuid.UUID
    user_id: Optional[uuid.UUID] = None
    status: str
    requested_sections: List[str]
    source_count: int
    records_found: int
    records_validated: int
    records_rejected: int
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class ResearchRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    business_id: uuid.UUID
    research_job_id: Optional[uuid.UUID] = None
    source_url: Optional[str] = None
    source_trust: str
    research_type: str
    field_name: str
    raw_value: Optional[str] = None
    normalized_value: str
    confidence: str
    evidence_text: Optional[str] = None
    observed_at: datetime
    expires_at: Optional[datetime] = None
    status: str
    meta_info: Optional[Dict[str, Any]] = None
    created_at: datetime


class ResearchConflictResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    business_id: uuid.UUID
    field_name: str
    competing_values: List[Dict[str, Any]]
    status: str
    resolution_notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class BusinessResearchProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    business_id: uuid.UUID
    business_name: str
    total_jobs: int
    total_records: int
    active_conflicts_count: int
    last_researched_at: Optional[datetime] = None
    confidence_score: int  # 0 to 100 percentage
    records: List[ResearchRecordResponse] = Field(default_factory=list)
    conflicts: List[ResearchConflictResponse] = Field(default_factory=list)
